File: 201415/S152.pas
var
  x, y, l: array[1..2000] of longint;
  n, m, k, ans, i, j, lc, left: longint;
begin
  read(n, m, k);
  for i := 1 to k do
    read(x[i], y[i]);
  ans := 1;
  for i := 1 to k do
  begin
    lc := 0;
    for j := 1 to k do
      if ((y[j] >= y[i]) and (y[j] < y[i] + m)) then
      begin
        inc(lc);
        l[lc] := x[j]
      end;
    left := 1;
    for j := 2 to lc do
    begin
      while (l[left] <= l[j] - m) do
        inc(left);
      if (j - left + 1 > ans) then
        ans := j - left + 1
    end
  end;
  writeln(ans)
end.

File: 201415/S152.cpp
#include <cstdio>
int x[2000], y[2000], l[2000];
int main() {
  int n, m, k;
  scanf("%d %d %d", &n, &m, &k);
  for (int i = 0; i < k; i++) {
    scanf("%d %d", &x[i], &y[i]);
  }
  int ans = 1;
  for (int i = 0; i < k; i++) {
    int lc = 0;
    for (int j = 0; j < k; j++) {
      if (y[j] >= y[i] && y[j] < y[i] + m) {
        l[lc++] = x[j];
      }
    }
    int left = 0;
    for (int j = 1; j < lc; j++) {
      while (l[left] <= l[j] - m) {
        left++;
      }
      if (j - left + 1 > ans) {
        ans = j - left + 1;
      }
    }
  }
  printf("%d\n", ans);
  return 0;
}
